terraform/nvflare/iam.tf

```diff
@@ -0,0 +1,23 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+module "storage_bucket_iam_bindings" {
+  source          = "terraform-google-modules/iam/google//modules/storage_buckets_iam"
+  version         = "7.7.1"
+  storage_buckets = [module.buckets.bucket.name]
```

Comment:
Use `var.workspace_bucket_name` instead of `module.buckets.bucket.name` otherwise this depends on apply, and fails if `module.buckets` doesn't exist. Example: when running `terraform destroy`

Reply:
Until this is fixed :(

https://github.com/terraform-google-modules/terraform-google-iam/issues/92

Reply:
I tried to change to var.workspace_bucket_name but as I'm generating a unique bucket, I miss the random suffix generated, therefore it can't work if I refactor it :(

Reply:
I think we need to fix this because it's going to break in several places. For example, even a `terraform plan` might report this error.

The main issue is that the `terraform-google-modules/iam/google//modules/storage_buckets_iam` submodule (as any other submodule in the `terraform-google-modules/iam` module) expects that the `storage_buckets` input variable is known before running `terraform apply`, which might make sense from the module point of view.

Even if you used the `random_id` resource (or any other `random_` resource) yourself to compute a random suffix to append to the bucket name and use that suffixed name, you'd the same issue because that value is still going to be known only after apply. (this is true when not targeting specific resources when running Terraform.)

Some options, in order of implementation effort:

- Use resource targeting when running Terraform. Quick to implement, but leaks logic outside Terraform because you have to remember (or script!) this.
- Require the storage name as an input variable, set `randomize_suffix` to `false` (the default) when creating the bucket, and update the README to say that you also need to set that (maybe with a command to generate a random suffix using Bash)
- Split the provisioning process in stages by moving this and other resources in another Terraform module that you invoke separately. This is probably a more structured approach, but has challenges as well, such as multiple Terraform states to manage.

I'd recommend the second option. It's probably good enough for an example.